### PR TITLE
feat: preserve ignore_index in semantic segmentation training

### LINKED ISSUE #765: Ignore index functionality
Description
I’d like to propose support for an ignore_index in the semantic segmentation training workflow, especially in geoai.train_segmentation_model() and SemanticSegmentationDataset.

At the moment, SemanticSegmentationDataset.__getitem__() normalizes label masks like this:

if self.num_classes == 2:
    if unique_vals.max() > 1:
        label_mask = (label_mask > 0).astype(np.int64)
else:
    label_mask = np.clip(label_mask, 0, self.num_classes - 1)

For multiclass segmentation, this means any ignored/no-data label such as PyTorch’s default -100 is clipped to class 0. Those pixels are then trained as background rather than ignored. This can significantly affect training when masks contain large no-data/ignore regions, which is common in geospatial segmentation workflows.

Expected outcome:

Add an ignore_index parameter to train_segmentation_model(), defaulting to -100 or None.
Pass it into SemanticSegmentationDataset.
Preserve ignored pixels during label normalization and resizing.
Use torch.nn.CrossEntropyLoss(ignore_index=ignore_index) when training.
Pass ignore_index into semantic segmentation metrics so validation metrics also exclude ignored pixels.
This would make GeoAI much safer for multiclass geospatial segmentation datasets where labels include no-data, masked-out areas, or unlabeled pixels.

Source code
Here is an example of the kind of change I’m proposing:

class SemanticSegmentationDataset(Dataset):
    def __init__(
        self,
        image_paths,
        label_paths,
        transforms=None,
        num_channels=None,
        target_size=None,
        resize_mode="resize",
        num_classes=2,
        ignore_index=-100,
    ):
        self.image_paths = image_paths
        self.label_paths = label_paths
        self.transforms = transforms
        self.target_size = target_size
        self.resize_mode = resize_mode
        self.num_classes = num_classes
        self.ignore_index = ignore_index
        ...
Then in __getitem__():

ignore_mask = None
if self.ignore_index is not None:
    ignore_mask = label_mask == self.ignore_index

unique_vals = np.unique(label_mask)

if self.num_classes == 2:
    if unique_vals.max() > 1:
        label_mask = (label_mask > 0).astype(np.int64)
else:
    label_mask = np.clip(label_mask, 0, self.num_classes - 1)

if ignore_mask is not None:
    label_mask[ignore_mask] = self.ignore_index
And in train_segmentation_model():

criterion = torch.nn.CrossEntropyLoss(ignore_index=ignore_index)

The resize path should also preserve ignored labels. For example, masks can still be resized with nearest-neighbor interpolation, but the result should not be clamped afterward in a way that converts ignore_index into a valid class.

I'm happy to add this feature myself if you'd like.

### PR DESCRIPTION
**Pull request by giswqs**
## Summary
- Adds an `ignore_index` parameter (default `-100`) to `SemanticSegmentationDataset` and `train_segmentation_model()`, threading it through label normalization, resizing, the default `CrossEntropyLoss`, and the validation metrics (`f1_score`, `iou_coefficient`, `precision_score`, `recall_score`, `evaluate_semantic`).
- Previously, no-data/ignored pixels in multi-class masks were silently clipped to class `0` (or normalized to foreground in the binary case), so PyTorch's default `-100` sentinel — or any custom no-data value — was being trained as a real class. This polluted both training and validation on geospatial datasets with masked-out regions.
- Passing `ignore_index=None` is supported and fully disables the ignore-index path (label normalization, loss, metrics), preserving the prior behavior for users who want it.

Closes #765

## Test plan
- [x] `pytest tests/test_train_ignore_index.py --verbose` (5 new tests, all pass)
  - Multi-class normalization preserves a negative `ignore_index` (`-100`) instead of clipping to class `0`.
  - Binary normalization preserves a positive sentinel (e.g. `255`) instead of mapping it to foreground.
  - Nearest-neighbor resizing keeps ignored pixels ignored.
  - `f1_score` / `iou_coefficient` / `precision_score` / `recall_score` exclude ignored pixels from scoring.
  - Metrics return `0.0` (rather than erroring) when every target pixel is ignored.
- [x] `pre-commit run --files geoai/train.py tests/test_train_ignore_index.py` (black, codespell, bandit, etc. — all pass).

### CODE REVIEW

File: geoai/train.py [hunk 1]
```diff
@@ -3247,15 +3286,15 @@ def _resize_image_and_mask(
...
+            mask = self._clamp_mask_tensor(mask)
 
         elif self.resize_mode == "pad":
             # Pad to target size (preserves aspect ratio)
             image = self._pad_to_size(image, (target_h, target_w))
             mask = self._pad_to_size(mask.unsqueeze(0), (target_h, target_w)).squeeze(0)
-            # Clamp mask values to ensure they're within valid range [0, num_classes-1]
-            mask = torch.clamp(mask, 0, self.num_classes - 1)
+            # Clamp valid mask values without converting ignore_index to a class.
+            mask = self._clamp_mask_tensor(mask)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply geoai/train.py]: Done in 3e8a989f — `_pad_to_size` now takes a `pad_value`, and `_resize_image_and_mask` passes `pad_value=self.ignore_index` for masks when set, so padded pixels stay ignored.

File: geoai/train.py [hunk 2]
```diff
@@ -3247,15 +3286,15 @@ def _resize_image_and_mask(
                 .squeeze(0)
                 .long()
             )
-            # Clamp mask values to ensure they're within valid range [0, num_classes-1]
-            mask = torch.clamp(mask, 0, self.num_classes - 1)
+            # Clamp valid mask values without converting ignore_index to a class.
+            mask = self._clamp_mask_tensor(mask)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply geoai/train.py:3292]: Done in 3e8a989f — annotations updated to `torch.Tensor` for both arguments and the return tuple.

File: tests/test_train_ignore_index.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    def test_metrics_return_zero_when_all_pixels_ignored(self):
+        """All-ignored targets should produce a defined zero score."""
+        pred = torch.tensor([[0, 1], [1, 0]])
+        target = torch.full((2, 2), -100, dtype=torch.long)
+
+        self.assertEqual(
+            iou_coefficient(pred, target, num_classes=2, ignore_index=-100),
+            0.0,
+        )
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/test_train_ignore_index.py:117]: Done in 3e8a989f — `test_metrics_return_zero_when_all_pixels_ignored` now asserts `f1_score`, `precision_score`, and `recall_score` also return `0.0` for all-ignored targets.
